src/wp-includes/template.php

```diff
@@ -698,9 +698,17 @@ function get_attachment_template() {
  * @return string The template filename if one is located.
  */
 function locate_template( $template_names, $load = false, $load_once = true, $args = array() ) {
-	$stylesheet_path = get_stylesheet_directory();
-	$template_path   = get_template_directory();
-	$is_child_theme  = $stylesheet_path !== $template_path;
+	global $stylesheet_path, $template_path;
+
+	if ( ! isset( $stylesheet_path ) ) {
+		$stylesheet_path = get_stylesheet_directory();
+	}
+
+	if ( ! isset( $template_path ) ) {
+		$template_path = get_template_directory();
+	}
```

Comment:
This memoization makes me worried. We already had memoization for this internally in `get_stylesheet_directory()` and `get_template_directory()`, which got reverted in https://github.com/WordPress/wordpress-develop/commit/b6bf3553d975da2e017a00443b7639aacf0a8586 due to a stream of weird edge-cases.

While not exactly the same, this code effectively reinstantes the memoization for when calling `locate_template()`. Maybe this is fine, but I'm a bit wary about it given the previous problems of the very similar idea.

Given that the ticket is about fixing the problem of using the initial site's directories even when switched on a multisite, it would be safer to avoid the memoization IMO, which partly is a separate performance consideration. We should only access the previously stored values from the initial site when being in a switched state.

The other problem even when doing that would be: What if the first time `locate_template()` is called happens while switched to another site? Then this would store the wrong data. I think to make this reliable we should store the output of these functions in the globals at a point that is determinate in the WordPress bootstrap process, some time earlier when the current theme is loaded and we can be sure we're in the initial site's state (i.e. not switched to another site).

Assuming the initial site's directories get stored in `$wp_stylesheet_path` and `$wp_template_path` globals, then this code here could be something like:
```suggestion
	global $wp_stylesheet_path, $wp_template_path;

	if ( is_multisite() && ms_is_switched() ) {
		$stylesheet_path = $wp_stylesheet_path;
		$template_path   = $wp_template_path;
	} else {
		$stylesheet_path = get_stylesheet_directory();
		$template_path   = get_template_directory();
	}
```

This way there's no memoization, except when in a switched state, in which case there never was.